### PR TITLE
core/state, core, miner: handle missing root error from state.New

### PR DESCRIPTION
**Pull request by Gustav-Simonsson**
This turned into
qdvd216

After a lot of discussion and feedback from @fjl this should now be ready for review again:
- [x] Bubble up error from state.New if missing state root
- [x] If this error is hit in miner, short-circuit `commitNewWork`
- [x] In tx pool, delay init of pendingState until new block or tx arrives.

### CODE REVIEW

**Comment by robotally** [comment]
| Vote | Count | Reviewers |
| :---: | :---: | :---: |
| :+1: | 0 |  |
| :-1: | 0 |  |

_Updated: Tue Oct 6 14:36:59 UTC 2015_

File: core/chain_makers.go [hunk 1]
```diff
@@ -155,8 +155,11 @@ func (b *BlockGen) OffsetTime(seconds int64) {
 // Blocks created by GenerateChain do not contain valid proof of work
 // values. Inserting them into BlockChain requires use of FakePow or
 // a similar non-validating proof of work implementation.
-func GenerateChain(parent *types.Block, db ethdb.Database, n int, gen func(int, *BlockGen)) []*types.Block {
-	statedb := state.New(parent.Root(), db)
+func GenerateChain(parent *types.Block, db ethdb.Database, n int, gen func(int, *BlockGen)) ([]*types.Block, error) {
+	statedb, err := state.New(parent.Root(), db)
+	if err != nil {
+		return nil, err
```
Comment (fjl) [inline core/chain_makers.go]: You can panic here. GenerateChain is not supposed to fail.

File: core/state/statedb.go [hunk 2]
```diff
@@ -291,8 +290,11 @@ func (self *StateDB) CreateAccount(addr common.Address) vm.Account {
 // Setting, copying of the state methods
 //
 
-func (self *StateDB) Copy() *StateDB {
-	state := New(common.Hash{}, self.db)
+func (self *StateDB) Copy() (*StateDB, error) {
+	state, err := New(common.Hash{}, self.db)
+	if err != nil {
+		return nil, err
+	}
```
Comment (fjl) [inline core/state/statedb.go]: Most of the other edits are cause by this change. Copy doesn't need to return an error because 
the state that is being copied already exists. The implementation of Copy uses New because it initializes
all the struct fields.

File: rpc/api/debug.go [hunk 3]
```diff
@@ -119,9 +119,9 @@ func (self *debugApi) DumpBlock(req *shared.Request) (interface{}, error) {
 		return nil, fmt.Errorf("block #%d not found", args.BlockNumber)
 	}
 
-	stateDb := state.New(block.Root(), self.ethereum.ChainDb())
-	if stateDb == nil {
-		return nil, nil
+	stateDb, err := state.New(block.Root(), self.ethereum.ChainDb())
+	if stateDb == nil || err != nil {
```
Comment (fjl) [inline rpc/api/debug.go]: It won't be nil if the error is.

**Comment by obscuren** [comment]
Are you looking for a `try { } catch { }` @Gustav-Simonsson :-P

**Comment by obscuren** [comment]
This has turned in to quite a mess. Is this required **just** for tests or are there requirements for normal operation as well? It works fine currently without the need to return errors.

If this is just for tests I'll suggestion something nasty: `panic` with a `recover` :-D

**Comment by Gustav-Simonsson** [comment]
@obscuren hehe, just wanted to see how big the change would be if naively bubbling up this error, as start of discussion of how to fix it :) will see how small the changeset becomes if changing state.Copy per @fjl suggestion. I can't say how likely this error is during normal operation, @fjl knows more there since he build the new trie.

**Comment by fjl** [comment]
`trie.New` now returns an error if the root node is missing from the database.
This PR is supposed to ensure that a missing root node causes chain insertion to fail.
It's not just for tests

File: xeth/xeth.go [hunk 4]
```diff
@@ -207,14 +211,21 @@ func (self *XEth) RemoteMining() *miner.RemoteAgent { return self.agent }
...
+			st, err = state.New(block.Root(), self.backend.ChainDb())
+			if err != nil {
+				return nil
+			}
 		} else {
-			st = state.New(self.backend.BlockChain().GetBlockByNumber(0).Root(), self.backend.ChainDb())
+			st, err = state.New(self.backend.BlockChain().GetBlockByNumber(0).Root(), self.backend.ChainDb())
+			if err != nil {
+				return nil
+			}
```
Comment (fjl) [inline xeth/xeth.go]: @obscuren This is one of the places where this change is actually useful.
Consider a full node that is at block 33000 but only has state roots starting at block 30000.
If someone requests a state value from block 20000 (for which the node doesn't have the state)
you'd rather make that visible somehow and return a proper error.

File: core/transaction_pool.go [hunk 5]
```diff
@@ -71,7 +71,12 @@ type TxPool struct {
 	queue   map[common.Address]map[common.Hash]*types.Transaction
 }
 
-func NewTxPool(eventMux *event.TypeMux, currentStateFn stateFn, gasLimitFn func() *big.Int) *TxPool {
+func NewTxPool(eventMux *event.TypeMux, currentStateFn stateFn, gasLimitFn func() *big.Int) (*TxPool, error) {
+	currentState, err := currentStateFn()
+	if err != nil {
+		return nil, err
+	}
```
Comment (fjl) [inline core/transaction_pool.go]: This is an unfortunate change. Might be better to delay the initialization of `pendingState` until the first call to
`resetState` or `checkQueue`.

Comment (Gustav-Simonsson) [inline core/transaction_pool.go]: Why is it unfortunate? It shouldn't change any behaviour of the pool, since if it happens it will simply refuse to start the pool.

Comment (fjl) [inline core/transaction_pool.go]: Because the pool doesn't actually need the state until txs are added, by which time the root might be available. Creating the pool shouldn't fail.

File: core/state/managed_state.go [hunk 6]
```diff
@@ -38,8 +38,9 @@ type ManagedState struct {
 
 // ManagedState returns a new managed state with the statedb as it's backing layer
 func ManageState(statedb *StateDB) *ManagedState {
+	copy := statedb.Copy()
 	return &ManagedState{
-		StateDB:  statedb.Copy(),
+		StateDB:  copy,
 		accounts: make(map[string]*account),
```
Comment (fjl) [inline core/state/managed_state.go]: Please revert this change.

File: core/genesis.go [hunk 7]
```diff
@@ -60,7 +60,10 @@ func WriteGenesisBlock(chainDb ethdb.Database, reader io.Reader) (*types.Block,
 		return nil, err
 	}
 
-	statedb := state.New(common.Hash{}, chainDb)
+	statedb, err := state.New(common.Hash{}, chainDb)
+	if err != nil {
+		return nil, err
+	}
```
Comment (fjl) [inline core/genesis.go]: The error can't happen here because creating with `common.Hash{}` always works.

File: core/genesis.go [hunk 8]
```diff
@@ -112,9 +115,12 @@ func WriteGenesisBlock(chainDb ethdb.Database, reader io.Reader) (*types.Block,
...
 
 // GenesisBlockForTesting creates a block in which addr has the given wei balance.
-// The state trie of the block is written to db.
+// The state trie of the block is written to db. the passed db needs to contain a state root
 func GenesisBlockForTesting(db ethdb.Database, addr common.Address, balance *big.Int) *types.Block {
-	statedb := state.New(common.Hash{}, db)
+	statedb, err := state.New(common.Hash{}, db)
+	if err != nil {
+		panic(err)
+	}
```
Comment (fjl) [inline core/genesis.go]: Same here, can't happen.

File: core/transaction_pool.go [hunk 9]
```diff
@@ -109,7 +114,17 @@ func (pool *TxPool) eventLoop() {
...
 func (pool *TxPool) resetState() {
-	pool.pendingState = state.ManageState(pool.currentState())
+	currentState, err := pool.currentState()
+	if err != nil {
+		glog.V(logger.Info).Infoln("Transaction pool: failed to get current state: %v", err)
+		return
+	}
+	managedState := state.ManageState(currentState)
+	if err != nil {
+		glog.V(logger.Info).Infoln("Transaction pool: failed to get managed state: %v", err)
```
Comment (fjl) [inline core/transaction_pool.go]: btw, the "Transaction pool: " prefix doesn't add much value because the log will say

```
timstamp  pid   txpool.go:124] message
```

File: core/transaction_pool.go (hunk 5 above)
Comment (fjl) [inline core/transaction_pool.go]: Looking at it another way: the txpool calls `currentStateFn` many times throughout its lifecycle,
any of these calls can fail with a missing root error. Why should the first call prevent creating the pool?

Comment (Gustav-Simonsson) [inline core/transaction_pool.go]: Can this actually fail though - ```currentStateFn``` is currently always ```eth.blockchain.State``` which should always succeed, right? It can only fail if ```blockchain``` has a inconsistent state where it's current block's state root is not present in it's db.

Comment (fjl) [inline core/transaction_pool.go]: Yeah, you're kind of right. Maybe txpool should keep a reference to BlockChain then (tight coupling FTW).

But this was the error that raised our attention, right? The state wasn't there when executing block tests.

**Comment by fjl** [comment]
:+1:
